20220714-lookup-api.md

```diff
@@ -0,0 +1,457 @@
+
+- **What is the expected outcome?**
+  OpenFGA clients use the Lookup API to return a list of object ids and then the client uses the list to filter and order their object list in their applications (pagination on the filtered list is the clients responsibility). The expectation is that the client consumes the whole Lookup response before filtering client side, and does not try to partially filter based on a subset of the response from Lookup. Hence, it is not the goal of Lookup to be able to continue the Lookup process from a prior invocation.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The Lookup API will provide clients with the ability to query the object ids that a user has a particular relationship with. The list of object ids the Lookup API returns can then be used by a client to filter an object collection to only the objects that a user has a particular relationship with by taking the intersection of it with any client provided filters/ordering of the underlying dataset. This is described in more detail with [Search with Permission (Option 3)](https://openfga.dev/docs/interacting/search-with-permissions#option-3-build-a-list-of-ids-then-search) document.
+
+It will be provided as two different API endpoints: one will provide a streaming endpoint and the other will implement the endpoint without streaming. These two different endpoints could be used interchangably, but clients can choose which one may be better for their integration environment.
+
+Consider a developer/client integrating with OpenFGA with the following authorization model:
+```
+type folder
+    relations
+        define viewer as self
+
+type document
+    relations
+        define viewer as self or editor or viewer from parent
+        define editor as self
+        define parent as self
+```
+and the following relationship tuples:
+```
+document:doc1, viewer, bob
+document:doc2, editor, bob
+document:doc3, parent, folder:folder1
+folder:folder1, viewer, bob
+```
+
+## StreamedLookup
+StreamedLookup will implement a grpc [server streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc) endpoint.
+
+The endpoint will stream each object id the target user has the given relationship with as it is evaluated, and the server will continuing streaming results until either the lookup has been exhausted or the `streamed-lookup-deadline` has been reached (see [Configuration Changes](#configuration-changes)).
+
+Sample grpcurl command:
+```
+grpcurl -plaintext \
+  -d '{
+           "store_id":"<store>", 
+           "authorization_model_id":"<modelID>",
+           "object_type":"document", 
+           "relation":"viewer", 
+           "user":"bob"
+      }' \
+  localhost:8081 openfga.v1.OpenFGAService/StreamedLookup
+
+{
+    "object_id": "doc1",
```

Comment:
For the response, what do you think about returning:
{
  "object": "$type:$id"
}

We don't use the object_id in any other interface

Reply:
If the client just wants the object identifier then they'll have to split the string every time 🤔 . That doesn't sound ideal.

The primary use case for Lookup is for pre-filtering on small object collections. If you have to iterate over all of othe object strings and split them to accumulate the object ids, then that's not a great client experience.

Reply:
Perhaps `type_id` or  simply `id` might be better. `


Agree that splitting isn't the most ideal, specially since we are not returning `tuple_key` here.

Reply:
Our documentation refers to "objects" all over the place.
https://openfga.dev/docs/concepts#what-is-an-object
https://openfga.dev/docs/modeling/getting-started#02-list-the-object-types
etc..

I think, to be consistent with our vernacular, that it makes sense to use `objectId` or `object_id`. 

`type_id` is odd, because it's not clear if it refers to the identifier of the type or the identifier of the object. `id` is a bit ambiguous, but it's  less ambiguous than `type_id` I think 🤷

Reply:
@rhamzeh this can get messier. I can't remember the reason for type rename etc but we have to have consistency in the concept names.

Reply:
I think returning  `"object_id": "doc1"` is better than returning `"object_id": "document:doc1"`

Reply:
What I was suggesting elsewhere in that document, is just returning the entire tuple, not just the ID.

This allows us in the future if we want to support querying for multiple relations or to have reverse lookup to use the same response semantics.

So the response will always be of the form `{ tuple_keys: TupleKey[] }` again consistent with our APIs elsewhere.

Reply:
If we are set on leaving it as is, `object_id` is fine by me.